Source/FluentSpreadsheets/Sheets/VisitorImplementations/SheetSegmentVisitor.cs

```diff
@@ -0,0 +1,174 @@
+    {
+        var headerVisitor = new HeaderSegmentVisitor<THeaderData>(_headerData);
+        var rowVisitor = new RowSegmentVisitor<THeaderData, TRowData>(_headerData, _rowData);
+        var footerVisitor = new FooterSegmentVisitor<THeaderData, TFooterData>(_headerData, _footerData);
+
+        builder.Accept(headerVisitor);
+        builder.Accept(rowVisitor);
+        builder.Accept(footerVisitor);
+
+        var header = headerVisitor.Component;
+        var row = rowVisitor.Components;
+        var footer = footerVisitor.Component;
+
+        return new SheetSegmentModel(header, row, footer);
+    }
+
+    private static void ScalePrototypedComponents<TDestination>(
+        IReadOnlyCollection<TDestination> data,
+        IList<IComponent> headerComponents,
+        IList<IList<IComponent>> rowComponents,
+        IList<IComponent> footerComponents)
+    {
+        for (var i = 0; i < data.Count; i++)
+        {
+            var rowWidth = CountRowWidth(i, headerComponents, rowComponents, footerComponents);
+
+            if (headerComponents.Count is not 0)
+                headerComponents[i] = ScaledComponent(headerComponents[i], rowWidth);
+
+            if (rowComponents.Count is not 0)
+            {
+                IList<IComponent>? segment = rowComponents[i];
+
+                for (var j = 0; j < segment.Count; j++)
+                {
+                    segment[j] = ScaledComponent(segment[j], rowWidth);
+                }
+            }
+
+            if (footerComponents.Count is not 0)
+                footerComponents[i] = ScaledComponent(footerComponents[i], rowWidth);
+        }
+    }
+
+    private static IComponent ScaledComponent(IComponent component, int width)
+    {
+        var scaleFactor = width / component.Size.Width;
+        return scaleFactor is 1 ? component : new ScaledComponent(component, new Scale(scaleFactor, 1));
+    }
+
+    private static int CountRowWidth(
+        int i,
+        IList<IComponent> headers,
+        IList<IList<IComponent>> rows,
+        IList<IComponent> footers)
+    {
+        IEnumerable<int> rowWidthEnumerable = Enumerable.Empty<int>();
+
+        if (headers.Count is not 0)
+            rowWidthEnumerable = rowWidthEnumerable.Append(headers[i].Size.Width);
+
+        if (rows.Count is not 0)
+            rowWidthEnumerable = rowWidthEnumerable.Concat(rows[i].Select(x => x.Size.Width));
+
+        if (footers.Count is not 0)
+            rowWidthEnumerable = rowWidthEnumerable.Append(footers[i].Size.Width);
+
+        var rowWidth = LcmCounter.Count(rowWidthEnumerable);
+        return rowWidth;
+    }
+
+    private static IEnumerable<IEnumerable<IComponent>> ConvertRowsFromSegmentGroupingToRowGrouping(
+        IList<IList<IComponent>> rows,
+        int length)
+    {
+        for (var i = 0; i < length; i++)
+        {
+            var localI = i;
+            yield return rows.Select(x => x[localI]);
+        }
```

Comment:
What's the point of having localI?

Reply:
It is the process called closure. As you can see, the variable is captured inside a lambda. All operations are executing lazily in that case, it means that enumeration probably will happen after the for-loop is done. If we capture `i`, all lambdas will get final value of it, by copying it local variable and capturing it, we make sure, that lambdas will receive proper values for their iterations. 

You can learn more about closures here:

https://youtu.be/h3MsnBRqzcY

Reply:
I knew a bit about what closures are, but never met them in real code, thanks for the explanation